optimum/onnxruntime/modeling_ort.py

```diff
@@ -632,21 +635,32 @@ def _output_shape_inference(self, axis_name: Union[str, int], dimensions: Dict[s
         """
         if isinstance(axis_name, int):
             return axis_name
-        result = dimensions.get(axis_name, None)
-        if result is None:
-            subparts = axis_name.split("+")
-            result = 0
-            for subname in subparts:
-                subname = subname.strip()
-                subresult = dimensions.get(subname, None)
-                if subresult is None:
-                    try:
-                        subresult = int(subname)
-                    except ValueError:
-                        result = None
-                        break
-                result += subresult
-        return axis_name if result is None else result
+        # It is actually covered below, but this is to make things faster.
+        elif axis_name in dimensions:
+            return dimensions[axis_name]
+
+        tokens = []
+        for idx, match in enumerate(re.finditer(self.output_shape_inference_pattern, axis_name)):
+            groups = match.groups()
+            matched_group = None
+            for idx, group in enumerate(groups):
+                if group is not None:
+                    matched_group = idx
+                    break
+            if matched_group is None:
+                raise ValueError(f"Could not perform shape inference on {axis_name}")
+
+            if matched_group == 0:
+                dim = dimensions.get(groups[0], None)
+                if dim is None or not isinstance(dim, int):
+                    return axis_name
+                tokens.append(str(dim))
+            else:
+                tokens.append(groups[matched_group])
+
+        # Here it should not be problematic to use eval since anything not matching the pattern would trigger an
```

Comment:
Rather than raising an exception, it's just not captured by `re.finditer(self.output_shape_inference_pattern, axis_name)`, right?

Reply:
No, if it's not captured it means the axis name contains something that is not supported, we want to fail here.

Reply:
Yes - so basically we enter into the for loop only if the expression matches with the regex, in which case we can expect the `eval()` not to fail, nor to run arbitrary code?

Reply:
Exactly we want to fail if nothing matches to not run arbitrary code.